src/pool.rs

```diff
@@ -494,55 +523,65 @@ impl ConnectionPool {
         guard[address.shard].remove(address);
     }
 
-    /// Check if a replica can serve traffic. If all replicas are banned,
-    /// we unban all of them. Better to try then not to.
-    pub fn is_banned(&self, address: &Address, role: Option<Role>) -> bool {
-        let replicas_available = match role {
-            Some(Role::Replica) => self.addresses[address.shard]
-                .iter()
-                .filter(|addr| addr.role == Role::Replica)
-                .count(),
-            None => self.addresses[address.shard].len(),
-            Some(Role::Primary) => return false, // Primary cannot be banned.
-        };
+    /// Check if address is banned
+    /// true if banned, false otherwise
+    pub fn is_banned(&self, address: &Address) -> bool {
+        let guard = self.banlist.read();
 
-        debug!("Available targets for {:?}: {}", role, replicas_available);
+        match guard[address.shard].get(address) {
+            Some(_) => true,
+            None => false,
+        }
+    }
 
-        let guard = self.banlist.read();
+    /// Determines if we can try to unban this server
+    pub async fn can_unban(&self, address: &Address) -> bool {
+        // If somehow primary ends up being banned we should return true here
+        if address.role == Role::Primary {
+            return true;
+        }
+
+        // Check if all instances are banned, in that case unban everything
+        let replicas_available = self.addresses[address.shard]
+            .iter()
+            .filter(|addr| addr.role == Role::Replica)
+            .count();
+
+        debug!("Available targets: {}", replicas_available);
 
-        // Everything is banned = nothing is banned.
-        if guard[address.shard].len() == replicas_available {
-            drop(guard);
-            let mut guard = self.banlist.write();
-            guard[address.shard].clear();
-            drop(guard);
+        let read_guard = self.banlist.read();
+        let all_replicas_banned = read_guard[address.shard].len() == replicas_available;
+        drop(read_guard);
+
+        if all_replicas_banned {
+            let mut write_guard = self.banlist.write();
             warn!("Unbanning all replicas.");
-            return false;
+            write_guard[address.shard].clear();
+            drop(write_guard);
```

Comment:
I drop it before issuing a `warn!` because `warn` is technically IO, so it's pretty slow. In this case though, it may be okay to wait 0.0001s it takes to print something to the screen before unlocking the mutex.